### PR TITLE
fix: parse the process name to get the process type instead of using port count

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: https://github.com/longhorn/longhorn/issues/7393

Get the process type by parsing the name instead of using port count.

### CODE REVIEW

**Comment by ChanYiLin** [comment]
Need to wait until IM PR merged, so I can update the vendor here.

File: engineapi/instance_manager.go [hunk 1]
```diff
@@ -315,19 +316,27 @@ func getDataEngineFromInstanceProcess(p *imapi.Instance) longhorn.DataEngineType
...
-func getTypeForInstance(instanceType longhorn.InstanceType, portCount int32) longhorn.InstanceType {
+func getTypeForInstance(instanceType longhorn.InstanceType, name string) longhorn.InstanceType {
 	if instanceType != longhorn.InstanceType("") {
 		return instanceType
 	}
 
-	if portCount == DefaultEnginePortCount {
+	// engine process name example: pvc-5a8ee916-5989-46c6-bafc-ddbf7c802499-e-0
+	nameSlices := strings.Split(name, "-")
+	processType := nameSlices[len(nameSlices)-2]
```
Comment (innobead) [inline engineapi/instance_manager.go:326]: Suggest using a regular expression to check instead. We can explicitly understand what the pattern is.

File: engineapi/instance_manager.go [hunk 2]
```diff
@@ -315,19 +316,27 @@ func getDataEngineFromInstanceProcess(p *imapi.Instance) longhorn.DataEngineType
 	return longhorn.DataEngineType(p.BackendStoreDriver)
 }
 
-func getTypeForInstance(instanceType longhorn.InstanceType, portCount int32) longhorn.InstanceType {
+func getTypeForInstance(instanceType longhorn.InstanceType, name string) longhorn.InstanceType {
 	if instanceType != longhorn.InstanceType("") {
```
Comment (innobead) [inline engineapi/instance_manager.go:320]: Can we have a `longhorn.InstanceNone` to replace the empty string here?

File: engineapi/instance_manager.go [hunk 3]
```diff
@@ -315,19 +316,27 @@ func getDataEngineFromInstanceProcess(p *imapi.Instance) longhorn.DataEngineType
...
+
+	if processType == "e" {
 		return longhorn.InstanceTypeEngine
 	}
 	return longhorn.InstanceTypeReplica
 }
 
-func getTypeForProcess(portCount int32) longhorn.InstanceType {
-	if portCount == DefaultEnginePortCount {
+func getTypeForProcess(name string) longhorn.InstanceType {
```
Comment (innobead) [inline engineapi/instance_manager.go:334]: Can just call `getTypeForInstance(longhorn.InstanceNone, name)` in this function to prevent duplicated code.

**Review by innobead — DISMISSED** [review]

File: engineapi/instance_manager.go (hunk 2 above)
Comment (ChanYiLin) [reply engineapi/instance_manager.go]: Fixed, thanks

File: engineapi/instance_manager.go (hunk 1 above)
Comment (ChanYiLin) [reply engineapi/instance_manager.go:326]: Fixed, thanks

File: engineapi/instance_manager.go (hunk 3 above)
Comment (ChanYiLin) [reply engineapi/instance_manager.go:334]: Fixed, thanks

**Comment by ChanYiLin** [comment]
Wait for https://github.com/longhorn/go-common-libs/pull/25

**Comment by innobead** [comment]
@ChanYiLin any updates?

**Comment by ChanYiLin** [comment]
@innobead
wait for https://github.com/longhorn/longhorn-instance-manager/pull/407

**Comment by ChanYiLin** [comment]
> This pull request is now in conflict. Could you fix it @ChanYiLin? 🙏

fixed.

**Comment by ChanYiLin** [comment]
> @ChanYiLin any updates?

Waiting for https://github.com/longhorn/longhorn-instance-manager/pull/407
So we can update the vendor in manager

File: engineapi/instance_manager.go [hunk 4]
```diff
@@ -15,6 +15,7 @@ import (
 	immeta "github.com/longhorn/longhorn-instance-manager/pkg/meta"
 	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
 
+	lhutils "github.com/longhorn/go-common-libs/utils"
```
Comment (derekbit) [inline engineapi/instance_manager.go:18]:
```
import (
	"context"
	"fmt"
	"path/filepath"
	"strconv"

	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
	"go.uber.org/multierr"

	imapi "github.com/longhorn/longhorn-instance-manager/pkg/api"
	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
	immeta "github.com/longhorn/longhorn-instance-manager/pkg/meta"
	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
	lhutils "github.com/longhorn/go-common-libs/utils"

	"github.com/longhorn/longhorn-manager/datastore"
	"github.com/longhorn/longhorn-manager/types"

	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
)
```

Comment (ChanYiLin) [reply engineapi/instance_manager.go]: @derekbit 
Fixed, thanks!
BTW, can you briefly explain what is our current import placement order?
Or is there any reference I can check, thanks

Comment (derekbit) [reply engineapi/instance_manager.go]: I learnt it from previous commits as well.

```
import (
	// golang standard library packages such as "fmt", "path/filepath" and so on.
	"context"
	"fmt"
	"path/filepath"
	"strconv"

        // third-party packages not related to k8s
	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
	"go.uber.org/multierr"

        // third-party packages without alias names related to k8s
        corev1 "k8s.io/api/core/v1"

        // third-party packages with alias names related to k8s
        "k8s.io/kubernetes/pkg/controller"

        // packages without alias names from longhorn components except for this repo 

        // packages with alias names from longhorn components except for this repo 
	imapi "github.com/longhorn/longhorn-instance-manager/pkg/api"
	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
	immeta "github.com/longhorn/longhorn-instance-manager/pkg/meta"
	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
	lhutils "github.com/longhorn/go-common-libs/utils"

        // packages without alias names from this repo
	"github.com/longhorn/longhorn-manager/datastore"
	"github.com/longhorn/longhorn-manager/types"

        // packages with alias names from this repo
        longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
)
```

**Review by derekbit — APPROVED** [review]

File: engineapi/instance_manager.go (hunk 4 above)
Comment (ChanYiLin) [reply engineapi/instance_manager.go]: Thanks!

**Pull request merged**